src/hotspot/share/prims/vectorSupport.cpp

```diff
@@ -277,19 +277,10 @@ InstanceKlass* VectorSupport::get_vector_payload_klass(BasicType elem_bt, int nu
 }
 
 Handle VectorSupport::allocate_vector_payload(InstanceKlass* ik, int num_elem, BasicType elem_bt, frame* fr, RegisterMap* reg_map, ObjectValue* ov, TRAPS) {
-  ScopeValue* payload = ov->field_at(0);
   intptr_t is_larval = StackValue::create_stack_value(fr, reg_map, ov->is_larval())->get_int();
   jint larval = (jint)*((jint*)&is_larval);
-
-  if (payload->is_location()) {
-    // Vector payload value in an aligned adjacent tuple (8, 16, 32 or 64 bytes).
-    return allocate_vector_payload_helper(ik, num_elem, elem_bt, larval, THREAD); // safepoint
-  } else if (!payload->is_object() && !payload->is_constant_oop()) {
-    stringStream ss;
-    payload->print_on(&ss);
-    assert(false, "expected 'object' value for scalar-replaced boxed vector but got: %s", ss.freeze());
-  }
-  return Handle(THREAD, nullptr);
+  // Vector payload value in an aligned adjacent tuple (8, 16, 32 or 64 bytes).
+  return allocate_vector_payload_helper(ik, num_elem, elem_bt, larval, THREAD); // safepoint
```

Comment:
I checked method `allocate_vector_payload_helper()`, and I didn't find any fields assignment inside it. Is this right or any further plan to add those assgnment? Besides, I have created a similar fix for the deoptimization: https://github.com/openjdk/valhalla/pull/863. Could you please take a look at it? Maybe we can integrate that change with this merge, or create a separate one if necessary.

Reply:
Oh, I see the followed calling to `Deoptimization::reassign_fields_by_klass()`. An alternative way is to keep the vector assignment in `vectorSupport.cpp` like before, and other fields reassign in `deoptimization.cpp` like other normal objects. So that we doesn't need to modify too much during deoptimization. The only thing we have to do is passing the secondary fields to safepoint if not vectorized. WDYT?

Reply:
> Oh, I see the followed calling to `Deoptimization::reassign_fields_by_klass()`. An alternative way is to keep the vector assignment in `vectorSupport.cpp` like before, and other fields reassign in `deoptimization.cpp` like other normal objects. So that we doesn't need to modify too much during deoptimization. The only thing we have to do is passing the secondary fields to safepoint if not vectorized. WDYT?

I think is better to completely de-couple allocation and re-assignment so that we have clean interface, having one routine for allocation and re-initialization worked until now because VectorBox could never have a scalar / constant inputs, but this is now relaxed since InlineTypeNodes created during fallback implementation can have non-vector inputs, thus a split in initialization b/w allocation and existing re-assignment routines looks clumsy to me,

Reply:
OK, make sense to me. BTW, do you think the condition `EnableVectorSupport` applied on the vector and vector_payload_mf classes in `Deoptimize::reasign_fields()` [1] and `Deoptimize::realloc_objects()` [2] can be removed? 

[1] https://github.com/openjdk/valhalla/blob/lworld%2Bvector/src/hotspot/share/runtime/deoptimization.cpp#L1548
[2] https://github.com/openjdk/valhalla/blob/lworld%2Bvector/src/hotspot/share/runtime/deoptimization.cpp#L1144

Reply:
> OK, I took another look at the changes in deoptimization.cpp, and the solution makes sense to me. Thanks!
> 
> BTW, do you think the condition `EnableVectorSupport` applied on the vector and vector_payload_mf classes in `Deoptimize::reasign_fields()` [1] and `Deoptimize::realloc_objects()` [2] can be removed now? Before, we only have the VectorBox cases which depends on this flag, but now, the default implementation also creates vector or scalar ops.
> 
> [1] https://github.com/openjdk/valhalla/blob/lworld%2Bvector/src/hotspot/share/runtime/deoptimization.cpp#L1548 [2] https://github.com/openjdk/valhalla/blob/lworld%2Bvector/src/hotspot/share/runtime/deoptimization.cpp#L1144

Thanks for suggesting this, ideally we can completely do away with special handling for vector box allocations since general case of instance allocations should suffice,  we are doing two additional things in vector allocations currently

1.  Larval bit setting in mark work of newly allocated oop.
2.  Explicit payload allocation and payload field initialization handling[1] for both flattened and non-flattened case,  but vector payloads always flatten with default InlineFieldMaxFlatSize limit. 

May be after the merge we can cleanup explicit allocation handling for vector boxes since re-assignment is already based on existing infra.

[1] https://github.com/jatin-bhateja/valhalla/blob/merge_lworld2/src/hotspot/share/prims/vectorSupport.cpp#L324